src/viewport-observer.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {devAssert} from '../src/log';
+import {isIframed} from './dom';
+
+/**
+ * Returns an IntersectionObserver tracking the Viewport.
+ * Only use this if x-origin iframe rootMargin support is considered nice-to-have,
+ * since if not supported this InOb will fallback to one without rootMargin.
+ *
+ * - If iframed: rootMargin is ignored unless natively supported (Chrome 81+).
+ * - If not iframed: all features work properly in both polyfill and built-in.
+ *
+ * @param {function(!Array<!IntersectionObserverEntry>)} ioCallback
+ * @param {!Window} win
+ * @param {number=|!Array<number>=} threshold
+ *
+ * @return {!IntersectionObserver}
+ */
+export function createViewportObserver(ioCallback, win, threshold) {
+  const iframed = isIframed(win);
+  const root = /** @type {?Element} */ (iframed
+    ? /** @type {*} */ (win.document)
+    : null);
+
+  // TODO(#30794): See if we can safely remove rootMargin without adversely
+  // affecting metrics.
+  const rootMargin = '25%';
+
+  try {
+    return new win.IntersectionObserver(ioCallback, {
```

Comment:
Why is this try catch necessary?

Reply:
Same reason as here. I can copy/paste the comment since its likely future viewers will have the same question

https://github.com/ampproject/amphtml/blob/6418ea956dd8cec31e68e945a33dfa37012d06d7/src/service/loading-indicator.js#L63-L65

Reply:
In Safari:
<img width="409" alt="Screen Shot 2020-10-21 at 7 13 38 PM" src="https://user-images.githubusercontent.com/4656974/96799637-942cf900-13d1-11eb-9846-c9d256b8443d.png">